### PR TITLE
Add a stream serializer

### PR DESCRIPTION
**Pull request by Matthew-Whitlock**
Enables serializing to/from an open stream.

### CODE REVIEW

File: src/checkpoint/serializers/stream_serializer.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
+//@HEADER
+// *****************************************************************************
+//
+//                             memory_serializer.h
```
Comment (PhilMiller) [inline src/checkpoint/serializers/stream_serializer.h]: Header name needs to be corrected here.

**Comment by PhilMiller** [comment]
Please include some suitable unit tests covering the new serializer

The code all looks fine to me

File: src/checkpoint/serializers/stream_serializer.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+#include "checkpoint/serializers/base_serializer.h"
+
+//Used to always static_assert false without compilation issues.
+template<class T> struct always_false : std::false_type {};
+
+namespace checkpoint {
+
+template <typename StreamT, eSerializationMode mode>
+struct StreamSerializer : Serializer {
+  //This really doesn't expect to be initialized like this, disabling until someone yells at me.
```
Comment (nmm0) [inline src/checkpoint/serializers/stream_serializer.h]: You can remove dead code

**Review by nmm0 — COMMENTED** [review]
Guard your code with `#if __cplusplus > 201703L` to compile only for C++17

**Comment by Matthew-Whitlock** [comment]
Replaced with #329

**Pull request closed without merging**